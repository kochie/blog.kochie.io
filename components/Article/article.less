@import "../common.less";

.card {
    box-shadow: 8px 14px 38px rgba(39,44,49,.06),1px 3px 8px rgba(39,44,49,.03);
    border-radius: 10px;
    // margin-left: 5%;
    // margin-right: 5%;
    width: 95%;
    margin-left: auto;
    margin-right: auto;
    position: relative;
    top: -25vh;
    background-color: white;
    padding: 5%;

    @media (min-width: @MediumWidth) and (orientation: landscape) {
        padding: 50px;
    }

    @media (min-width: 950px) and (orientation: landscape) {
        width: 900px;
        // padding: 50px;
        // margin-left: 10%;
        // margin-right: 10%;
    }

    @media (min-width: @LargeWidth) and (orientation: landscape) {
        // padding: 50px;
        // margin-left: 20%;
        // margin-right: 20%;
    }
}

.container {
    position: relative;
}

.title {
    position: absolute;
    z-index: 100;
    left: 50%;
    transform: translate(-50%, 0);
    text-align: center;
    color: white;
    margin-top: 20vh;   
    border-radius: 10px;
    overflow: hidden;


    @media (min-width: @MediumWidth) and (orientation: landscape) {
    }
}

.glass {
    padding: 5px 15px;
    border-radius: 5px;
    background-color: rgba(0,0,0,0.4);
    background-size: 100vw 70vh;
    background-position: center;
    filter: blur(10px);
    overflow: hidden;
    width: 100%;
    height: 100%;
    position: absolute;
    // filter: blur(10px);
    // filter: url("data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg'><filter id='svgMask'><feGaussianBlur stdDeviation='10' /></filter></svg>#svgMask");
}

.text {
    position: relative;
    margin-top: 0px;
    margin-bottom: 0px;
    padding-top: 5px;
    padding-bottom: 5px;
    padding-right: 15px;
    padding-left: 15px;
}

.subText{
    font-size: 11px;
    // letter-spacing: 0.1rem;
    text-transform: uppercase;
    color: #717375;
}

.heading {
    margin-top: 10px;
    margin-bottom: 10px;
}

.details {
    display: flex;
    flex-direction: row;
    justify-content: space-between
}