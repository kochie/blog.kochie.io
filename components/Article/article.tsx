import { Image, Footer } from "..";
import { article } from "articles.json";

import React from 'react'
import Error from "next/error";
import dynamic from "next/dynamic";
import {Tag, Set} from 'fannypack'

import style from './article.less';

export default (article: article) => {
    const DynamicComponent = dynamic(() => import(`../../posts/${article.articleFile}.mdx`).catch((error: {message: string}) => {
        if (error.message === `Cannot find module './${article.articleFile}.mdx'`) {
            return () => (
                <Error statusCode={404} />
            )
        } else {
            throw error
        }
    }), {
        loading: () => <p>...</p>
    });

    return (
        <>
            <div style={{backgroundColor: 'black'}}>
                {/* <div className={style.title}>
                    <div className={style.glass} style={{backgroundImage: `url(${article.jumbotron.lqip})`}}/>
                    <h1 className={style.text}></h1>
                </div> */}
                <Image height={'70vh'} width={'100vw'} {...article.jumbotron} />
            </div>
            <div className={style.container}>
            <div className={style.card}>
                <div>
                    <div><span className={style.subText}>
                        Published on {new Date(article.publishedDate).toLocaleDateString('en')}
                    </span></div>
                    <h1 className={style.heading}>{article.title}</h1>
                    <div className={style.details}>
                        <span className={style.subText}>Written by {article.author}</span>
                        {
                            !article.editedDate ? 
                            null : 
                            <span className={style.subText}>
                                Last edited on {new Date(article.editedDate).toLocaleDateString('en')}
                            </span>
                        }
                        <span className={style.subText}>{article.readTime} min read</span>
                    </div>
                    <Set style={{marginTop: '5px'}}>
                        {article.tags.map(tag => (<Tag kind="outlined">{tag}</Tag>))}
                    </Set>
                </div>
                <DynamicComponent />
            </div>
            </div>
            <Footer title={"Kochie Engineering"} links={[]}/>
        </>
    )
}