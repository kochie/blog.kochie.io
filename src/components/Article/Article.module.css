.title {
  position: absolute;
  z-index: 100;
  left: 50%;
  transform: translate(-50%, 0);
  text-align: center;
  color: white;
  margin-top: 20vh;
  border-radius: 10px;
  overflow: hidden;
}

.glass {
  padding: 5px 15px;
  border-radius: 5px;
  background-color: rgba(0, 0, 0, 0.4);
  background-size: 100vw 70vh;
  background-position: center;
  filter: blur(10px);
  overflow: hidden;
  width: 100%;
  height: 100%;
  position: absolute;
}

.text {
  position: relative;
  margin-top: 0px;
  margin-bottom: 0px;
  padding-top: 5px;
  padding-bottom: 5px;
  padding-right: 15px;
  padding-left: 15px;
}

.subText {
  font-size: 12px;
  text-transform: uppercase;
  color: var(--subtext-color);
}

.heading {
  margin-top: 10px;
  margin-bottom: 10px;
}

.details {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
}

.imageForeground {
  /* background-color: var(--image-foreground-background-color); */
  height: 100%;
  width: 100%;
  overflow: hidden;
}

.tagset {
  margin-left: -0.25em;
  margin-bottom: 10px;
}

.hr {
  width: 100px;
}

.underline {
  --animation: ease-in-out 0.25s; 
  position: relative;
  color: var(--heading-color);
  text-decoration: none;
  max-width: max-content;
  margin: 0;
  transition: var(--animation);
}

.underline:hover {
  color: var(--text-color);
  cursor: pointer;
}
.underline:before {
  content: "";
  position: absolute;
  width: 100%;
  height: 2px;
  bottom: 0;
  left: 0;
  background-color: var(--heading-color);
  visibility: hidden;
  transform: scaleX(0);
  transition: var(--animation);
}

.underline:hover:before {
  visibility: visible;
  transform: scaleX(1);
  background-color: var(--text-color);
}