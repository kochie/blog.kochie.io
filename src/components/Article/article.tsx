import { Image, Jumbotron } from '..'
import { Article as ArticleMetadata } from 'articles.json'
import { Author as AuthorMetadata } from 'authors.json'

import React from 'react'
// import Error from 'next/error'
// import dynamic from 'next/dynamic'

import style from '../../styles/article.module.scss'
import Link from 'next/link'
import Card from '../Card'
import { Tag, TagSet } from '../Tag'

interface ArticleProps {
  article: ArticleMetadata
  author: AuthorMetadata
  children: React.ReactNode
  jumbotron: Image
}

// export async function getStaticProps(context) {

// }

const Article = ({ article, author, children, jumbotron }: ArticleProps) => {
  // const DynamicComponent = dynamic(
  //   () =>
  //     import(`../../../articles/${article.articleDir}/index.mdx`).catch(
  //       (error: { message: string }) => {
  //         if (
  //           error.message ===
  //           `Cannot find module './${article.articleDir}/index.mdx'`
  //         ) {
  //           return () => (
  //             <Error title={"Article doesn't exist"} statusCode={404} />
  //           )
  //         } else {
  //           throw error
  //         }
  //       }
  //     ),
  //   {
  //     loading: () => <p> ... </p>
  //   }
  // )

  const AuthorLink = () => (
    <Link href={'/authors/[authorId]'} as={`/authors/${author.username}`}>
      <a className={style.underline}>{author.fullName}</a>
    </Link>
  )

  return (
    <>
      <Jumbotron
        height={'70vh'}
        width={'100vw'}
        background={
          <div style={{ width: '100vw', height: '70vh' }}>
            <Image src={jumbotron.url} lqip={jumbotron.lqip} />
          </div>
        }
        foreground={<div className={style.imageForeground} />}
        // foreground={
        // <div className={style.title}>
        //             <div className={style.glass} style={{backgroundImage: `url(${article.jumbotron.lqip})`}}/>
        //             <h1 className={style.text}></h1>
        //         </div>
        // }
      />
      <div className={style.container}>
        <div>
          <Card>
            <div className={style.card}>
              <h1 className={style.heading}>{article.title}</h1>
              <TagSet className={style.tagset}>
                {article.tags.map((tag) => (
                  // <div key={tag}>{tag}</div>
                  <Tag name={tag} link={`/tags/${tag}`} key={tag} />
                ))}
              </TagSet>
              {/* <hr className={style.hr} /> */}
              <div>
                <span className={style.subText}>
                  {`Published on ${new Date(
                    article.publishedDate
                  ).toLocaleDateString('en')}`}
                </span>
              </div>
              <div className={style.details}>
                <div>
                  <div>
                    <span className={style.subText}>
                      {`Written by `}
                      <AuthorLink />
                    </span>
                  </div>
                  <div>
                    {!article.editedDate ? null : (
                      <span className={style.subText}>
                        {`Last edited on ${new Date(
                          article.editedDate
                        ).toLocaleDateString('en')}`}
                      </span>
                    )}
                  </div>
                  <div>
                    <span className={style.subText}>
                      {article.readTime} min read
                    </span>
                  </div>
                </div>
              </div>
              <div style={{ marginTop: '5px' }}></div>
              {/* <DynamicComponent /> */}
              {children}
            </div>
          </Card>
        </div>
      </div>
      {/* <Footer title={'Kochie Engineering'} links={[]} /> */}
    </>
  )
}

export default Article
