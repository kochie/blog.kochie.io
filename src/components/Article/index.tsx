import React, { PropsWithChildren } from 'react'
import Link from 'next/link'
import Jumbotron from '@/components/Jumbotron'
import Card from '@/components/Card'
import { Tag, TagSet } from '@/components/Tag'
import Image from 'next/image'

// eslint-disable-next-line import/no-unresolved
// import { Article as ArticleDetails } from 'articles.json'
// eslint-disable-next-line import/no-unresolved
// import { Author as AuthorDetails } from 'authors.json'

import style from './Article.module.css'
import { ArticleMetadata } from 'src/lib/article-path'
import { Author } from 'metadata.yaml'

interface ArticleProps {
  article: ArticleMetadata
  author: Author
}

interface AuthorLinkProps {
  username: string
  fullname: string
}

const Article = ({
  article,
  author,
  children,
}: PropsWithChildren<ArticleProps>): React.ReactElement => {
  const AuthorLink = ({
    username,
    fullname,
  }: AuthorLinkProps): React.ReactElement => {
    const link = (
      <Link href={'/authors/[authorId]'} as={`/authors/${username}`}>
        <a className={style.underline}>{fullname}</a>
      </Link>
    )
    return (
      <>
        <p>Written by {link}</p>
      </>
    )
  }

  // console.log(decodeBlurHash(article.jumbotron.lqip))

  return (
    <>
      <Jumbotron
        height={'60vh'}
        width={'100vw'}
        background={
          <div className="">
            <Image
              src={article.jumbotron.url}
              layout="fill"
              blurDataURL={article.jumbotron.lqip}
              objectFit="cover"
              objectPosition="center"
              placeholder="blur"
            />
          </div>
        }
        foreground={<div className="h-full w-full overflow-hidden" />}
      />
      <div className="relative max-w-5xl -mt-20 mx-auto px-4 mb-0 pb-10">
        <div>
          <Card>
            <div className="p-4 md:p-8 lg:p-14">
              <h1 className="my-3 text-3xl md:text-left text-center">
                {article.title}
              </h1>
              <TagSet className="md:mb-3 md:-ml-1 justify-center md:justify-start">
                {article.tags.map((tag) => (
                  <Tag name={tag} link={`/tags/${tag}`} key={tag} inverted />
                ))}
              </TagSet>
              <div className="flex flex-col md:flex-row justify-between items-center">
                <div>
                  <span className={style.subText}>
                    {`Published on ${new Date(
                      article.publishedDate
                    ).toLocaleDateString('en')}`}
                  </span>
                </div>
                <div>
                  {article.editedDate == article.publishedDate ? null : (
                    <span className={style.subText}>
                      {`Last edited on ${new Date(
                        article.editedDate
                      ).toLocaleDateString('en')}`}
                    </span>
                  )}
                </div>
              </div>
              <div className="flex flex-col md:flex-row justify-between items-center">
                <span className={style.subText}>
                  <AuthorLink
                    username={author.username}
                    fullname={author.fullName}
                  />
                </span>
                <div>
                  <span className={style.subText}>{article.readTime}</span>
                </div>
              </div>
              <div style={{ marginTop: '5px' }}></div>
              {children}
            </div>
          </Card>
        </div>
      </div>
    </>
  )
}

export default Article
