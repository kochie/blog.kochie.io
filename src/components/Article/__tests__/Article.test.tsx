import React from 'react'
import { ReactTestRenderer, act, create } from 'react-test-renderer'
import { IconProp } from '@fortawesome/fontawesome-svg-core'
import Article, { AuthorLink } from '@/components/Article'


const testArticle = {
  title: 'title',
  author: 'author',
  blurb: 'blurb',
  jumbotron: {
    url: '/test.png',
    alt: 'alt text',
    lqip: 'AAAAAAAAAAAA',
  },
  articleDir: 'articleDir',
  readTime: '1 min read',
  tags: ['some', 'tags'],
  publishedDate: '2019-06-27T10:59:18.365Z',
  editedDate: '2019-06-28T10:59:18.365Z',
  indexPath: '',
  path: '',
}

const icon: IconProp = 'accessible-icon'

const testAuthor = {
  username: 'string',
  fullName: 'string',
  email: 'string',
  socialMedia: [
    {
      name: 'string',
      link: 'string',
      icon: icon,
      color: 'string',
      tracking: 'tracking',
    },
  ],
  avatar: {
    src: 'string',
    lqip: 'AAAAAAAAAAAA',
  },
  bio: 'string',
}

const TestArticle = (
  <div>
    <p>This is a test article</p>
  </div>
)

describe('ARTICLE COMPONENT', () => {
  test('renders correctly', () => {
    let tree: ReactTestRenderer

    act(() => {
      tree = create(
        <Article article={testArticle} author={testAuthor}>
          {TestArticle}
        </Article>
      )
    })

    // @ts-expect-error tree will be assigned
    expect(tree.toJSON()).toMatchSnapshot()
  })
})

describe('AUTHORLINK COMPONENT', () => {
  test('renders correctly', () => {
    let tree: ReactTestRenderer

    act(() => {
      tree = create(<AuthorLink username={'username'} fullname={'fullname'} />)
    })

    // @ts-expect-error tree will be assigned
    expect(tree.toJSON()).toMatchSnapshot()
  })
})
