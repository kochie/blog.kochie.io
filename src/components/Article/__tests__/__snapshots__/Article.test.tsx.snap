// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ARTICLE COMPONENT renders correctly 1`] = `
[
  <div
    className="bg-black overflow-hidden relative"
    style={
      {
        "height": "60vh",
        "width": "100vw",
      }
    }
  >
    <div
      className="relative filter brightness-75 blur-sm w-full h-full transform-gpu scale-110"
    >
      <div
        className="relative w-full h-[60vh]"
      >
        <img
          alt="alt text"
          data-nimg="fill"
          decoding="async"
          loading="lazy"
          onError={[Function]}
          onLoad={[Function]}
          sizes="100vw"
          src="/_next/image?url=%2Ftest.png&w=1200&q=75"
          srcSet="/_next/image?url=%2Ftest.png&w=320&q=75 320w, /_next/image?url=%2Ftest.png&w=420&q=75 420w, /_next/image?url=%2Ftest.png&w=768&q=75 768w, /_next/image?url=%2Ftest.png&w=1024&q=75 1024w, /_next/image?url=%2Ftest.png&w=1200&q=75 1200w"
          style={
            {
              "backgroundImage": "url("data:image/svg+xml;charset=utf-8,%3Csvg xmlns='http%3A//www.w3.org/2000/svg'%3E%3Cimage style='filter:blur(20px)' preserveAspectRatio='xMidYMid slice' x='0' y='0' height='100%25' width='100%25' href='AAAAAAAAAAAA'/%3E%3C/svg%3E")",
              "backgroundPosition": "center",
              "backgroundRepeat": "no-repeat",
              "backgroundSize": "cover",
              "bottom": 0,
              "color": "transparent",
              "height": "100%",
              "left": 0,
              "objectFit": "cover",
              "objectPosition": "center",
              "position": "absolute",
              "right": 0,
              "top": 0,
              "width": "100%",
            }
          }
        />
      </div>
    </div>
    <div
      className="relative -top-full w-full h-full"
    >
      <div
        className="h-full w-full overflow-hidden"
      />
    </div>
  </div>,
  <div
    className="relative max-w-5xl -mt-20 mx-auto px-4 mb-0 pb-10"
  >
    <div>
      <div
        className="shadow-2xl rounded-2xl transition ease-in-out duration-200 bg-gray-300 dark:bg-neutral-600 text-black dark:text-white w-full h-full z-20 dark:shadow-neutral-800"
      >
        <div
          className="p-4 md:p-8 lg:p-14"
        >
          <h1
            className="mt-3 mb-10 text-5xl md:text-left text-center"
          >
            title
          </h1>
          <div
            className="md:mb-3 md:-ml-1 justify-center md:justify-start tagSet"
          >
            <a
              href="/tags/some"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              <div
                className="tag-inverted tag-common"
              >
                some
              </div>
            </a>
            <a
              href="/tags/tags"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              <div
                className="tag-inverted tag-common"
              >
                tags
              </div>
            </a>
          </div>
          <div
            className="flex flex-col md:flex-row justify-between items-center uppercase text-xs"
          >
            <div>
              <span>
                Published on 
                <time
                  dateTime="2019-06-27T10:59:18.365Z"
                >
                  27/06/2019
                </time>
              </span>
            </div>
            <div>
              <span>
                Last edited on 
                <time
                  dateTime="2019-06-28T10:59:18.365Z"
                >
                  28/06/2019
                </time>
              </span>
            </div>
          </div>
          <div
            className="flex flex-col md:flex-row justify-between items-center uppercase text-xs"
          >
            <span>
              <p>
                Written by 
                <a
                  className="underline dark:text-gray-300"
                  href="/authors/string"
                  onClick={[Function]}
                  onMouseEnter={[Function]}
                  onTouchStart={[Function]}
                >
                  string
                </a>
              </p>
            </span>
            <div>
              <span>
                1 min read
              </span>
            </div>
          </div>
          <div
            style={
              {
                "marginTop": "5px",
              }
            }
          />
          <div>
            <p>
              This is a test article
            </p>
          </div>
        </div>
      </div>
      <div
        className="visible"
      >
        <div
          className="fixed bottom-6 right-6 h-20 w-20 bg-green-500 rounded-full z-10 group animate-bounce-out"
        >
          <div
            style={
              {
                "transform": "rotate(90deg) scaleX(-1)",
              }
            }
          >
            <svg
              viewBox="0 0 50 50"
            >
              <circle
                className="progress-circle"
                cx="25"
                cy="25"
                fill="transparent"
                r="22"
                stroke="red"
                strokeDasharray={138.16}
                strokeDashoffset={0}
                strokeWidth={6}
              />
            </svg>
          </div>
          <div
            className="fixed top-2 left-2 bg-white rounded-full h-16 w-16 shadow-2xl cursor-pointer"
            onClick={[Function]}
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-arrow-up-to-line fa-2x fa-stack-1x group-hover:animate-bounce-orig text-black"
              data-icon="arrow-up-to-line"
              data-prefix="fad"
              focusable="false"
              role="img"
              style={{}}
              viewBox="0 0 384 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                className="fa-duotone-group"
                style={{}}
              >
                <path
                  className="fa-secondary"
                  d="M0 64C0 46.3 14.3 32 32 32H352c17.7 0 32 14.3 32 32s-14.3 32-32 32H32C14.3 96 0 81.7 0 64z"
                  fill="currentColor"
                  style={{}}
                />
                <path
                  className="fa-primary"
                  d="M169.4 169.4c12.5-12.5 32.8-12.5 45.3 0l128 128c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L224 269.3V448c0 17.7-14.3 32-32 32s-32-14.3-32-32V269.3L86.6 342.6c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3l128-128z"
                  fill="currentColor"
                  style={{}}
                />
              </g>
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`AUTHORLINK COMPONENT renders correctly 1`] = `
<p>
  Written by 
  <a
    className="underline dark:text-gray-300"
    href="/authors/username"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onTouchStart={[Function]}
  >
    fullname
  </a>
</p>
`;
